### PR TITLE
add a new expression type: Correlated Column.

### PR DESCRIPTION
**Pull request by hanfei1991**
In past, we use a correlated flag to check if the column is correlated. Now we add a new expression type to represent it. This will make us free from checking the correlated flag everytime. Sometimes we will forget it.
@shenli @coocood @zimulala @XuHuaiyu @tiancaiamao PTAL

### CODE REVIEW

**Comment by coocood** [comment]
Please explain what's the benefit of this change.

**Comment by coocood** [comment]
LGTM

**Comment by hanfei1991** [comment]
@zimulala PTAL

**Comment by ngaut** [comment]
PTAL @XuHuaiyu

File: plan/logical_plan_builder.go [hunk 1]
```diff
@@ -911,54 +910,59 @@ type ApplyConditionChecker struct {
 	All       bool
 }
 
-func (b *planBuilder) buildApply(p, inner LogicalPlan, schema expression.Schema, checker *ApplyConditionChecker) LogicalPlan {
+func (b *planBuilder) buildApply(outerPlan, innerPlan LogicalPlan, checker *ApplyConditionChecker) LogicalPlan {
```
Comment (XuHuaiyu) [inline plan/logical_plan_builder.go]: I think that add simple comment for innerPlan and outerPlan will be more readable.

File: plan/logical_plans.go [hunk 2]
```diff
@@ -93,11 +93,11 @@ type Selection struct {
...
 
-	InnerPlan   LogicalPlan
-	OuterSchema expression.Schema
-	Checker     *ApplyConditionChecker
-	// outerColumns is the columns that not belong to this plan.
-	outerColumns []*expression.Column
+	InnerPlan        LogicalPlan
+	Checker          *ApplyConditionChecker
+	corColsInCurPlan []*expression.CorrelatedColumn
+	// outerColumns is the correlated columns that not belong to this plan.
```
Comment (XuHuaiyu) [inline plan/logical_plans.go]: 1. update this comment, no `outerColumns`
2. s/`not belong`/`don't belong`

**Comment by XuHuaiyu** [comment]
@hanfei1991

File: plan/expression_rewriter.go [hunk 3]
```diff
@@ -162,11 +157,11 @@ func (er *expressionRewriter) handleCompareSubquery(v *ast.CompareSubqueryExpr)
 		er.err = errors.Errorf("Unknown compare type %T.", v.R)
 		return v, true
 	}
-	np, outerSchema := er.buildSubquery(subq)
+	np := er.buildSubquery(subq)
 	if er.err != nil {
 		return v, true
 	}
-	// Only (a,b,c) = all (...) and (a,b,c) != any () can use row expression.
+	// Only (a,b,c) = all (...) and (a,b,c) != any () can use row
```
Comment (zimulala) [inline plan/expression_rewriter.go]: Add . at the end of a line, the same as line 231

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by zimulala** [comment]
LGTM

**Pull request merged**